### PR TITLE
TS Strict Collections

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: packages/@react-stately/collections/src/CollectionBuilder.ts [hunk 1]
```diff
@@ -183,7 +188,7 @@ export class CollectionBuilder<T extends object> {
     }
 
     // Ignore invalid elements
-    if (partialNode.key == null) {
+    if (partialNode.key == null || partialNode.type == null) {
```
Comment (LFDanLu) [inline packages/@react-stately/collections/src/CollectionBuilder.ts:191]: was `type` being defined always a implied requirement here? It makes sense, but just wanna double check

File: packages/@react-types/shared/src/collections.d.ts [hunk 2]
```diff
@@ -189,20 +189,22 @@ export interface Node<T> {
   /** A unique key for the node. */
   key: Key,
   /** The object value the node was created from. */
-  value: T | null,
+  value: T | null | undefined,
   /** The level of depth this node is at in the hierarchy. */
   level: number,
   /** Whether this item has children, even if not loaded yet. */
-  hasChildNodes: boolean,
+  hasChildNodes?: boolean,
```
Comment (LFDanLu) [inline packages/@react-types/shared/src/collections.d.ts:196]: guess this has always been technically optional due to partialNode's type?

File: packages/@react-stately/collections/src/CollectionBuilder.ts (hunk 1 above)
Comment (snowystinger) [reply packages/@react-stately/collections/src/CollectionBuilder.ts:191]: I think so? the type of the full node has these both as required, so I figure if we don't have them, then we have an invalid element

File: packages/@react-types/shared/src/collections.d.ts (hunk 2 above)
Comment (snowystinger) [reply packages/@react-types/shared/src/collections.d.ts:196]: it's been 4+ months, i have no recollection of why I decided to do this here as opposed to like the previous one and consider it to be an invalid element

**Review by LFDanLu — APPROVED** [review]
From what I can tell, those changes should be theoretically innocuous, 🤞

**Comment by snowystinger** [comment]
closing for https://github.com/adobe/react-spectrum/pull/7368

**Pull request closed without merging**